src/ls_odbc.c

```diff
@@ -601,32 +600,135 @@ static int create_connection (lua_State *L, int o, env_data *env, SQLHDBC hdbc)
+	SQLRETURN ret;
+
+	ret = SQLSetEnvAttr (env->henv, SQL_ATTR_ODBC_VERSION, (void *)SQL_OV_ODBC3, 0);
+	if (error(ret)) {
+		return luasql_faildirect (L, "error setting SQL version.");
+	}
+
+	/* tries to allocate connection handle */
+	ret = SQLAllocHandle (hDBC, env->henv, &hdbc);
+	if (error(ret)) {
+		return luasql_faildirect (L, "connection allocation error.");
+	}
+
+	/* tries to connect handle */
+	ret = SQLDriverConnect (hdbc, NULL, sourcename, SQL_NTS, sqlOutBuf, 4096,
+	                        &sqlOutLen, SQL_DRIVER_NOPROMPT);
+	if (error(ret)) {
+		ret = fail(L, hDBC, hdbc);
+		SQLFreeHandle(hDBC, hdbc);
+		return ret;
+	}
+
+	/* success, return connection object */
+	ret = create_connection (L, 1, env, hdbc);
+	if(ret == 1) {
+		/* Add the sqlOutBuf string to the results, for diagnostics */
+		lua_pushlstring(L, (char *)sqlOutBuf, sqlOutLen);
+		return 2;
+	}
+
+	return ret;
+}
 
 /*
-** Creates and returns a connection object
-** Lua Input: source [, user [, pass]]
+** Reforms old connection style to new one
+** Lua Input: source, [user, [pass]]
 **   source: data source
 **   user, pass: data source authentication information
 ** Lua Returns:
+**   new connection details table
+*/
+static void env_connect_fix_old (lua_State *L)
+{
+	static const char *const opt_names[] = {
+		"source",
+		"user",
+		"password",
+		NULL
+	};
+	int i, t = lua_gettop(L)-1;
+
+	lua_newtable(L);
+	lua_insert(L, 2);
+
+	for(i=0; opt_names[i] != NULL && i<t; ++i) {
+		lua_pushstring(L, opt_names[i]);
+		lua_pushvalue(L, i+3);
+		lua_settable(L, 2);
+	}
+
+	lua_settop(L, 2);
+}
+
+/*
+** Creates and returns a connection object
+** Lua Input: {
+**    source = <database source/path>,
+**    user = <user name>,
+**    password = <user password>,
+** or
+**    dsn = <DSN string>,
+** }
+** Lua Returns:
 **   connection object if successfull
 **   nil and error message otherwise.
 */
-static int env_connect (lua_State *L) {
-	env_data *env = (env_data *) getenvironment (L);
-	SQLCHAR *sourcename = (SQLCHAR*)luaL_checkstring (L, 2);
-	SQLCHAR *username = (SQLCHAR*)luaL_optstring (L, 3, NULL);
-	SQLCHAR *password = (SQLCHAR*)luaL_optstring (L, 4, NULL);
-	SQLHDBC hdbc;
-	SQLRETURN ret;
+static int env_connect (lua_State *L)
+{
+	env_data *env = getenvironment (L);
+	SQLCHAR *sourcename = NULL;
+	SQLCHAR *username = NULL;
+	SQLCHAR *password = NULL;
+	SQLHDBC hdbc = NULL;
+	SQLRETURN ret = 0;
+
+	if(lua_gettop(L) < 2) {
+		return luasql_faildirect(L, "No connection details provided");
+	}
+
+	if(!lua_istable(L, 2)) {
+		env_connect_fix_old(L);
+	}
+
+	/* check for the custom DSN connection string */
+	if(luasql_table_optstring(L, 2, "dsn", NULL) != NULL) {
+		return env_table_connect_DSN(L);
+	}
+
+	/* get the standard connection details */
+	sourcename = (SQLCHAR *)luasql_table_optstring(L, 2, "source", NULL);
+	username = (SQLCHAR *)luasql_table_optstring(L, 2, "user", NULL);
+	password = (SQLCHAR *)luasql_table_optstring(L, 2, "password", NULL);
```

Comment:
I'd suggest it provides an easy pattern for conversion of the other drivers to the more flexible named parameter style (which this patch shows a good case for)

There's likely going to be additional params added here at some point (e.g. #36) and giving a clear, single place the values are picked up (along with a clear place for backward compatibility) seems a logical arrangement. The function inlines and the overhead of a table is nothing compared to the actual job of connecting to a DB.